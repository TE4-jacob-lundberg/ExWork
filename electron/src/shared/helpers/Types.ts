export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  id: string;
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}
