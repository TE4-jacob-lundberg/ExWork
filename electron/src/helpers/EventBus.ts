/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
