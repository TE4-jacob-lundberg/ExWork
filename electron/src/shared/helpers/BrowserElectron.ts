export const electron = require('electron');
