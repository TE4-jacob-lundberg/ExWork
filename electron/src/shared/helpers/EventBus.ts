/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
