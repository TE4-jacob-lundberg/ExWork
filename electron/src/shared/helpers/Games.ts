import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
