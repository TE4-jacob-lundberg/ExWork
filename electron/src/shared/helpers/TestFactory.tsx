/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
