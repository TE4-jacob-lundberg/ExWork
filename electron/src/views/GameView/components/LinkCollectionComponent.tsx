import React, { useState } from 'react';
import Styled from '@emotion/styled';
import { keyframes } from '@emotion/core';

import { ILink } from '../../../shared/helpers/Types';
import LinkComponent from '../../../shared/components/LinkComponent';

interface Props {
  links: ILink[];
}

const loadedKeyframes = keyframes`
  0% { max-height: 0%; }
  100% { max-height: 100%; }
`;

const LinkCollectionComponent: React.FC<Props> = function (props: Props) {
  const [isAnimating, setIsAnimating] = useState(true);

  const ContainerStyled = Styled.div`
    background: #444;
    width: 200px;
    max-height: 100%;
    animation: ${isAnimating ? loadedKeyframes : 'none'} 1s ease-out;
    overflow: ${isAnimating ? 'hidden' : 'auto'};
    display: grid;
    grid-template-columns: 1fr;
    opacity: 0.8;

    &::-webkit-scrollbar {
      display: none;
    }
  `;

  return (
    <ContainerStyled onAnimationEnd={(): void => setIsAnimating(false)}>
      {props.links.map((link, i) => 
        <LinkComponent
          key={i}
          link={link}
          background={i % 2 ? '#777' : '#444'}
        />
      )}
    </ContainerStyled>
  );
};

export default LinkCollectionComponent;
