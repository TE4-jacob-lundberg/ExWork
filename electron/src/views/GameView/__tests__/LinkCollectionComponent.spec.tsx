import React from 'react';
import { factory } from '../../../shared/helpers/TestFactory';

import LinkCollectionComponent from '../components/LinkCollectionComponent';

const testComponent = <LinkCollectionComponent 
  links={[{url: '', label: 'link1'}, {url: '', label: 'link2'}]}
/>;

describe('LinkCollectionComponent.tsx', () => {
  it('renders correctly', () => {
    const { container } = factory(testComponent);
    
    expect(container).toMatchSnapshot();
  });

  it('displays the links', () => {
    const { getByText } = factory(testComponent);

    expect(getByText('link1')).toBeTruthy();
    expect(getByText('link2')).toBeTruthy();
  });
});
