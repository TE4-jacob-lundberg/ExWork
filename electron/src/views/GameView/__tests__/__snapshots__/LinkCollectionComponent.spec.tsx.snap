// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkCollectionComponent.tsx renders correctly 1`] = `
<div>
  <div
    class="css-cjmacl"
  >
    <div
      class="css-nijhdc"
    >
      <button
        class="css-1qbtfbw"
      >
        link1
      </button>
    </div>
    <div
      class="css-1svfgn6"
    >
      <button
        class="css-1qbtfbw"
      >
        link2
      </button>
    </div>
  </div>
</div>
`;
